File: src/db-gwent-server-scala/src/main/scala/com/entities/Card.scala
package com.entities

import com.basic.{Entity, t, removeIf}
import com.entities.abilities.Ability

case class CardId(value: String)
case class Card(
  id: CardId,
  name: String,
  $type: CardType,
  row: CardRowType,
  points: PointGroup = PointGroup(),
  abilities: Array[Ability]) extends Entity[CardId]:

  def hasWeatherAbility: Boolean =
    $type == CardType.Leader

  def pointsTo1(id: String = ""): Boolean =
    points.modifiers.addOne:
      PointGroupModifier(1, PointGroupModifierType.Constant, id)
    .isEmpty

  def removeModifier(id: String): Boolean =
    points.modifiers.removeIf:
      modifier => modifier.id == id
    .isDefined

File: src/db-gwent-server-kotlin/src/main/kotlin/com/entities/Card.kt
package com.entities

import com.entities.abilities.Ability

class Card(
    val id: CardId,
    val name: String,
    val type: CardType,
    val row: RowType,
    val points: PointGroup = PointGroup(),
    val abilities: Array<Ability>) {

    fun pointsTo1(id: String = ""): Boolean =
        points.modifiers.add(
            PointGroupModifier(
                value = 1, PointGroupModifierType.Constant, id) { points.originalValue })

    fun addOriginalPoints(id: String): Boolean =
        points.modifiers.add(
            PointGroupModifier(
                value = 1, PointGroupModifierType.MultiplyOriginal, id) { points.originalValue })

    fun multiplyPoints(value: Int, id: String): Boolean =
        points.modifiers.add(
            PointGroupModifier(
                value, PointGroupModifierType.Multiply, id) { points.originalValue })

    fun removeModifier(id: String): Boolean =
        points.modifiers.removeIf { it.id == id }
}

data class CardId(val value: String)
